src/app/fdctl/run/tiles/fd_poh.c

```diff
@@ -526,6 +529,107 @@ static fd_poh_ctx_t * fd_poh_global_ctx;
 static volatile ulong fd_poh_waiting_lock __attribute__((aligned(128UL)));
 static volatile ulong fd_poh_returned_lock __attribute__((aligned(128UL)));
 
+/* Agave also needs to write to some mcaches, so we trampoline
+   that via. the PoH tile as well. */
+
+struct poh_link {
+  fd_frag_meta_t * mcache;
+  ulong            depth;
+  ulong            tx_seq;
+
+  void *           mem;
+  void *           dcache;
+  ulong            chunk0;
+  ulong            wmark;
+  ulong            chunk;
+
+  ulong            cr_avail;
+  ulong            rx_cnt;
+  ulong *          rx_fseqs[ 32UL ];
+};
+
+typedef struct poh_link poh_link_t;
+
+poh_link_t gossip_dedup;
+poh_link_t stake_out;
+poh_link_t crds_shred;
+
+poh_link_t replay_plugin;
+poh_link_t gossip_plugin;
+poh_link_t start_progress_plugin;
+poh_link_t vote_listener_plugin;
+poh_link_t poh_plugin;
+
+static void
+poh_link_wait_credit( poh_link_t * link ) {
+  if( FD_LIKELY( link->cr_avail ) ) return;
+
+  while( 1 ) {
+    ulong cr_query = ULONG_MAX;
+    for( ulong i=0UL; i<link->rx_cnt; i++ ) {
+      ulong const * _rx_seq = link->rx_fseqs[ i ];
+      ulong rx_seq = FD_VOLATILE_CONST( *_rx_seq );
+      ulong rx_cr_query = (ulong)fd_long_max( (long)link->depth - fd_long_max( fd_seq_diff( link->tx_seq, rx_seq ), 0L ), 0L );
+      cr_query = fd_ulong_min( rx_cr_query, cr_query );
+    }
+    if( FD_LIKELY( cr_query>0UL ) ) {
+      link->cr_avail = cr_query;
+      break;
+    }
+    FD_SPIN_PAUSE();
+  }
+}
+
+static void
+poh_link_publish( poh_link_t *  link,
+                  ulong         sig,
+                  uchar const * data,
+                  ulong         data_len ) {
+  while( FD_UNLIKELY( !FD_VOLATILE_CONST( link->mcache ) ) ) FD_SPIN_PAUSE();
```

Comment:
I know this is not new code, but is it safe to assume that when the tile boots, `link->mcache` will be `NULL`? Will it not be uninitialized memory in a workspace?

Reply:
These are in bss, so they are zero initialized